packages/react/src/ReferenceInput/ReferenceInput.tsx

```diff
@@ -16,75 +17,253 @@ export interface ReferenceInputProps {
+    return { targetType, resourceType: targetType?.resourceType as ResourceType };
+  });
 
-  function setValueHelper(newValue: Reference | undefined): void {
-    setValue(newValue);
-    if (props.onChange) {
-      props.onChange(newValue);
+  const searchCriteria = useMemo<ReferenceInputProps['searchCriteria']>(() => {
+    if (state.targetType?.type === 'profile') {
+      return { ...props.searchCriteria, _profile: state.targetType.value };
     }
-  }
+    return props.searchCriteria;
+  }, [props.searchCriteria, state.targetType]);
+
+  useEffect(() => {
+    if (!targetTypes) {
+      return;
+    }
+
+    if (targetTypes.filter(shouldFetchResourceType).length === 0) {
+      return;
+    }
+
+    const newTargetTypePromises: Promise<TargetType>[] = targetTypes.map((tt) => {
+      if (shouldFetchResourceType(tt)) {
+        return fetchResourceTypeOfProfile(medplum, tt.value)
+          .then((profile) => {
+            const newTargetType = { ...tt };
+
+            if (!profile) {
+              console.debug(`StructureDefinition for ${tt.value} not found`);
+              newTargetType.error = 'StructureDefinition not found';
+            } else if (!profile.type || isEmpty(profile.type)) {
+              console.debug(`resourceType for ${tt.value} unexpectedly missing`);
+              newTargetType.error = 'resourceType unavailable';
+            } else {
+              newTargetType.resourceType = profile.type satisfies string;
+              newTargetType.name = profile.name;
+              newTargetType.title = profile.title;
+            }
+
+            return newTargetType;
+          })
+          .catch((reason) => {
+            console.error(reason);
+            return { ...tt, error: reason };
+          });
+      } else {
+        return Promise.resolve(tt);
+      }
+    });
+
+    Promise.all(newTargetTypePromises)
+      .then((newTargetTypes) => {
+        setTargetTypes(newTargetTypes);
+        if (state.targetType) {
+          const needle: string = state.targetType.value;
+          const index = newTargetTypes.findIndex((tt) => tt.value === needle);
+          if (index >= 0) {
+            // orphaned targetType has been resolved
+            setState({
+              ...state,
+              targetType: newTargetTypes[index],
+              resourceType: newTargetTypes[index].resourceType as ResourceType,
+            });
+          } else {
+            console.log(`defaultValue had unexpected resourceType: ${state.targetType.resourceType}`);
+          }
+        }
+      })
+      .catch(console.error);
+  }, [medplum, state, state.targetType, targetTypes]);
+
+  const setValueHelper = useCallback(
+    (item: Resource | undefined) => {
+      const newValue = item ? createReference(item) : undefined;
+      setValue(newValue);
+      if (props.onChange) {
+        props.onChange(newValue);
+      }
+    },
+    // exhaustive deps wants this to be [props] instead of [props.onChange], but it's
```

Comment:
My guess is that `props` could possibly be a new object on every render (like in the case of prop spreading) so they probably try to make you include `props` whenever you include `props.{{SOME_PROP}}`. In this case I think it's fine.

Reply:
Yeah, must be something like that; I splatted it to avoid the comment & eslint-disable per Cody's suggestion.